### PR TITLE
Avoid double-compacting data in bottom level in manual compactions

### PR DESCRIPTION
**Pull request by miasantreble**
Depending on the config, manual compaction (leveled compaction style) does following compactions:
L0->L1
L1->L2
...
Ln-1 -> Ln
Ln -> Ln
The final Ln -> Ln compaction is partly unnecessary as it recompacts all the files that were just generated by the Ln-1 -> Ln. We should avoid recompacting such files. This rule should be applied to Lmax only.
Resolves issue https://github.com/facebook/rocksdb/issues/4995

### CODE REVIEW

**Comment by miasantreble** [comment]
Hey @ajkr , do you mind taking a quick look to see if this approach makes sense? Thanks!

**Comment by ajkr** [comment]
Thanks for doing this, will take a look. BTW, do you mind reminding me what @yoshinorim's reason was for only operating on the bottom level? I remember he wrote some insightful reason on the task.

**Comment by ajkr** [comment]
Never mind, I think it makes sense. We wouldn't want to skip files in middle levels because the latest data for the range (as of the time `CompactRange` is called) might not be included by other compactions happening in parallel.

File: db/compaction_picker.cc [hunk 1]
```diff
@@ -660,6 +661,22 @@ Compaction* CompactionPicker::CompactRange(
...
+    std::vector<FileMetaData*> inputs_shrinked;
+    for (size_t i = 0; i < inputs.size(); ++i) {
+      if (inputs[i]->fd.GetNumber() <= max_sst_file_number) {
+        inputs_shrinked.emplace_back(inputs[i]);
+      }
+    }
+    if (inputs_shrinked.empty()) {
+      return nullptr;
+    }
+    inputs.files.swap(inputs_shrinked);
```
Comment (ajkr) [inline db/compaction_picker.cc]: Is it a problem that the files may not cover a contiguous key-range after this code filters some out?

I wonder if we should pick each compaction consisting of one contiguous range of files that all need to be compacted. We can then set `compaction_end` to the end of that range and another call will be made for the next contiguous range, etc.

**Comment by miasantreble** [comment]
@ajkr was in a meeting earlier, but here is from Yoshi:

>  For example:
Suppose calling CompactRange(), and max sst number was 100 at that time. Currently, compactions will happen like this:
 001.sst, 002.sst, 003.sst (L0) ->  101.sst (L1)
 004.sst, 005.sst, 101.sst, 006.sst (L1) -> 102.sst (L2)
 011.sst, 012.sst, 102.sst, 013.sst (L2) -> 103.sst (L3)
If you skip sst files > max number for all levels, several files (101.sst in L1 and 102.sst in L2 in this case) will remain in higher levels, which may waste some space. That's why I think only reasonable level to skip compaction is Lmax.

File: db/compaction_picker.cc (hunk 1 above)
Comment (ajkr) [reply db/compaction_picker.cc]: Hm, it looks like the following `ExpandInputsToCleanCut` refills any holes if we pick non-contiguous files here. It seems safe, but OTOH some cases may be suboptimal. For example, the first and last file might be the only ones that need compaction (i.e., `inputs_shrinked` contains only those two). But then `ExpandInputsToCleanCut` will repopulate the input files with everything in the range, which leaves us with no files skipped.

Comment (miasantreble) [reply db/compaction_picker.cc]: Thanks @ajkr !
let me see if I understood the idea correctly:
so in `CompactionPicker::CompactRange`, after the call to `GetOverlappingInputs`, we will still filter out new files larger than `max_sst_file_number`, then maybe call a new function to break up `inputs_shrinked` into continuous ranges of files, and for each range, call `SetupOtherInputs` and then create/register one new Compaction? 
This will create a number of `Compaction` objects but the caller `DBImpl::RunManualCompaction` seems to expect only one `Compaction`?

Comment (miasantreble) [reply db/compaction_picker.cc]: I guess if it's only suboptimal in some corner cases but optimizes in general cases, we can live with that. Or if really needed, we can also do some trick in `ExpandInputsToCleanCut` but that might be a bit error-prone

**Comment by ajkr** [comment]
I think you can return one compaction at a time consisting of a range of
contiguous files. The caller (manual compaction) will have to call
CompactRange again to get the next compaction of contiguous files. It can
know where to search for the next range of contiguous files if we set
compaction_end to the end of the current range.

**Comment by miasantreble** [comment]
I see what you mean now, `RunManualCompaction` will repeatedly call `manual.cfd->CompactRange` until manual.done is set to true, and every time `BackgroundCompaction` runs, it will update `begin` with `manual_end` so I guess all we need to do now is in CompactionPicker when filtering `inputs` by file number, stop as soon as we find a newly created file, let `ExpandInputsToCleanCut` do its job and run the compaction using the truncated inputs. Will try out this idea. Thanks!

File: db/compaction_picker.cc (hunk 1 above)
Comment (siying) [reply db/compaction_picker.cc]: Rather than leaving holes in the middle, should we just cut the compaction?
If we have files [a b] [c d] [e f], and [c d] is newly generated, should we pick [a b] to compact itself, and then [e f] in a following one?

**Comment by siying** [comment]
@miasantreble oops, I didn't see your latest comments. I was just trying to mention the same thing as you guys concluded:)

File: db/compaction_picker.cc [hunk 2]
```diff
@@ -660,6 +661,32 @@ Compaction* CompactionPicker::CompactRange(
   }
   assert(output_path_id < static_cast<uint32_t>(ioptions_.cf_paths.size()));
 
+  // for BOTTOM LEVEL compaction, skip files that are created during the
+  // current compaction
+  if (max_sst_file_number > 0) {
+    assert(input_level == output_level);
+    std::vector<FileMetaData*> inputs_shrinked;
```
Comment (siying) [inline db/compaction_picker.cc]: nit: shrink's passive term seems not to be "shrinked".

File: db/compaction_picker.cc [hunk 3]
```diff
@@ -660,6 +661,32 @@ Compaction* CompactionPicker::CompactRange(
...
+    for (size_t i = 0; i < inputs.size(); ++i) {
+      if (inputs[i]->fd.GetNumber() <= max_sst_file_number) {
+        inputs_shrinked.emplace_back(inputs[i]);
```
Comment (siying) [inline db/compaction_picker.cc]: push_back() is more straight-forward.

File: db/compaction_picker.cc [hunk 4]
```diff
@@ -660,6 +661,32 @@ Compaction* CompactionPicker::CompactRange(
...
+      } else {
+        // inputs[i] was created during the current manual compaction and
+        // need to be skipped
+        if (inputs_shrinked.empty()) {
+          // haven't found anything to compact, keep looking
+          continue;
+        } else {
+          // already found something to compact, skip the rest files
+          break;
```
Comment (siying) [inline db/compaction_picker.cc]: My personal prefence is that
```
else if (!inputs_shrinked.empty()) {
  break;
}
```
is easier to understand than the flow now.

File: db/compaction_picker.cc [hunk 5]
```diff
@@ -660,6 +661,32 @@ Compaction* CompactionPicker::CompactRange(
...
+        }
+      }
+    }
+    if (inputs_shrinked.empty()) {
+      return nullptr;
+    }
+    inputs.files.swap(inputs_shrinked);
+  }
```
Comment (siying) [inline db/compaction_picker.cc]: I think `covering_the_whole_range` needs to be updated accordingly.

File: db/db_compaction_test.cc [hunk 6]
```diff
@@ -4268,6 +4268,56 @@ TEST_F(DBCompactionTest, ManualCompactionFailsInReadOnlyMode) {
...
+  CompactRangeOptions cro;
+  cro.bottommost_level_compaction = BottommostLevelCompaction::kForce;
+  dbfull()->CompactRange(cro, nullptr, nullptr);
+
+  const std::vector<InternalStats::CompactionStats>& comp_stats2 =
+      internal_stats_ptr->TEST_GetCompactionStats();
+  num = comp_stats2[2].num_input_files_in_output_level;
+  ASSERT_EQ(num, 0);
+}
+
```
Comment (siying) [inline db/db_compaction_test.cc]: We have `compaction_picker_test.cc`. Can we have more thorough test coverage there for some corner cases?

File: db/db_impl_compaction_flush.cc [hunk 7]
```diff
@@ -639,6 +639,25 @@ void DBImpl::NotifyOnFlushCompleted(ColumnFamilyData* cfd,
...
+  FileType type;
+  for (size_t i = 0; i < files.size(); ++i) {
+    if (ParseFileName(files[i], &number, &type)) {
+      if (type == kTableFile && number > largest_file_number) {
+        largest_file_number = number;
+      }
+    }
+  }
+  return largest_file_number;
+}
```
Comment (siying) [inline db/db_impl_compaction_flush.cc]: Instead of doing this, can we just use VersionSet::current_next_file_number()?

File: db/db_impl_compaction_flush.cc [hunk 8]
```diff
@@ -731,6 +756,8 @@ Status DBImpl::CompactRange(const CompactRangeOptions& options,
           continue;
         }
         output_level = level;
+        // only use max_sst_file_number for bottom level compaction
+        max_sst_file_number_to_use = max_sst_file_number;
```
Comment (siying) [inline db/db_impl_compaction_flush.cc]: I think we should pick a better name than `max_sst_file_number_to_use ` and `max_sst_file_number`. I had a hard time understanding the logic.

File: db/column_family.cc [hunk 9]
```diff
@@ -984,11 +984,12 @@ Compaction* ColumnFamilyData::CompactRange(
     const MutableCFOptions& mutable_cf_options, int input_level,
     int output_level, uint32_t output_path_id, uint32_t max_subcompactions,
     const InternalKey* begin, const InternalKey* end,
-    InternalKey** compaction_end, bool* conflict) {
+    InternalKey** compaction_end, bool* conflict,
+    uint64_t max_sst_file_number) {
```
Comment (siying) [inline db/column_family.cc]: It's the same problem as many other places: the name `max_sst_file_number` is hard to make sense. I suggest we pick a better argument name here. Also, document what it means in comments.

**Comment by siying** [comment]
I gave it a pass, and I believe we are on the right track!

File: db/db_impl_compaction_flush.cc (hunk 8 above)
Comment (miasantreble) [reply db/db_impl_compaction_flush.cc]: renamed to `next_file_number` and also made default to `kMaxUint64` so hopefully it's easier to understand

File: db/column_family.cc [hunk 10]
```diff
@@ -985,11 +985,11 @@ Compaction* ColumnFamilyData::CompactRange(
     int output_level, uint32_t output_path_id, uint32_t max_subcompactions,
     const InternalKey* begin, const InternalKey* end,
     InternalKey** compaction_end, bool* conflict,
-    uint64_t max_sst_file_number) {
+    uint64_t next_file_number) {
```
Comment (siying) [inline db/column_family.cc]: Still not very easily understood to me. Imagine a user reads function CompactRange(), and sees a function named `next_file_number`, what he would think about it is?

Maybe something like `max_file_num_to_ignore`, or something like that?

File: db/compaction_picker.cc [hunk 11]
```diff
@@ -661,30 +661,29 @@ Compaction* CompactionPicker::CompactRange(
...
       }
     }
-    if (inputs_shrinked.empty()) {
+    if (inputs_shrunk.empty()) {
       return nullptr;
     }
-    inputs.files.swap(inputs_shrinked);
+    if (inputs.size() != inputs_shrunk.size()) {
+      inputs.files.swap(inputs_shrunk);
+      covering_the_whole_range = false;
```
Comment (siying) [inline db/compaction_picker.cc]: Why?

Comment (miasantreble) [reply db/compaction_picker.cc]: because `inputs` is shrunk here, so this compaction no longer covers the whole range. the check for `max_compaction_bytes` seems to be doing the same thing, no?

Comment (siying) [reply db/compaction_picker.cc]: Here is an example. Let's say we have three files: 1.sst, 2.sst, 3.sst. Initially we picked all of them, which make it cover the whole range. Now we eliminated 3 because it is new, but take [1, 2]. According to the logic here, we made `covering_the_whole_range`. With this, we'll update update compaction_end to be the largest key of 2.sst. And in the upper layer, it will call RunManualCompaction() again, just to find that there isn't anything to compact anymore. It's the same thing even if we eliminate 1.sst. It's not wrong but feels intuitive to me.

Comment (miasantreble) [reply db/compaction_picker.cc]: I see, so we should only set `covering_the_whole_range` to false when
1) we filtered out some "new" files in the for-loop
2) in vector `inputs` there are more than 1 "old" files yet to be compacted

File: db/db_compaction_test.cc [hunk 12]
```diff
@@ -4422,6 +4424,56 @@ TEST_F(DBCompactionTest, ManualCompactionFailsInReadOnlyMode) {
...
+      ASSERT_OK(Put("foo" + std::to_string(i*10+j), RandomString(&rnd, 1024)));
+    }
+    Flush();
+  }
+
+  MoveFilesToLevel(2);
+
+  for (auto i = 0; i < 8; ++i) {
+    for (auto j = 0; j < 10; ++j) {
+      ASSERT_OK(Put("bar" + std::to_string(i*10+j), RandomString(&rnd, 1024)));
```
Comment (siying) [inline db/db_compaction_test.cc]: make format

File: db/db_impl_compaction_flush.cc [hunk 13]
```diff
@@ -706,7 +712,9 @@ Status DBImpl::CompactRange(const CompactRangeOptions& options,
     }
     s = RunManualCompaction(cfd, ColumnFamilyData::kCompactAllLevels,
                             final_output_level, options.target_path_id,
-                            options.max_subcompactions, begin, end, exclusive);
+                            options.max_subcompactions,
+                            options.bottommost_level_compaction, begin, end,
```
Comment (siying) [inline db/db_impl_compaction_flush.cc]: We are passing three members of options in now. Do we consider to forward `const CompactRangeOptions&` instead?

Comment (miasantreble) [reply db/db_impl_compaction_flush.cc]: was thinking about that as well, will update the next upload

**Comment by miasantreble** [comment]
@ajkr do you mind taking another look before we land the PR? Thanks!

File: HISTORY.md [hunk 14]
```diff
@@ -18,6 +18,7 @@
 ## 6.1.0 (3/27/2019)
 ### New Features
 * Introduce two more stats levels, kExceptHistogramOrTimers and kExceptTimers.
+* Added BottommostLevelCompaction::kForceOptimized to avoid double compacting newly compacted files in bottom level compaction of manual compaction.
```
Comment (ajkr) [inline HISTORY.md]: For my understanding, is it an option because `kForce` should be able to force recompaction even on files that were trivially moved to the bottom level?

[the remark this answers is not in the document]
Comment (miasantreble) [reply HISTORY.md]: Right, thanks for pointing out. The protocol hasn't changed yet :)

Comment (miasantreble) [reply HISTORY.md]: Yes, we decided to make it an option because tools like `option_change_migration` and tests like `DBCompactionTestWithParam .ConvertCompactionStyle` expects manual compactions to generate one single file in the bottom level
